src/pip/_internal/resolution/resolvelib/factory.py

```diff
@@ -160,19 +164,35 @@ def _iter_found_candidates(
         if not ireqs:
             return ()
 
-        # The InstallRequirement implementation requires us to give it a
-        # "template". Here we just choose the first requirement to represent
-        # all of them.
-        # Hopefully the Project model can correct this mismatch in the future.
-        template = ireqs[0]
-        name = canonicalize_name(template.req.name)
-
         extras = frozenset()  # type: FrozenSet[str]
         for ireq in ireqs:
             specifier &= ireq.req.specifier
             hashes &= ireq.hashes(trust_internet=False)
             extras |= frozenset(ireq.extras)
 
+        # HACK: The InstallRequirement implementation requires us to give it a
+        # "template", so we try to "invent" one by aggregating available
+        # information from all the ireqs. This feels very brittle, hopefully
+        # the Project model can correct this mismatch in the future.
+        name = canonicalize_name(ireqs[0].req.name)
+        template = InstallRequirement(
+            PackagingRequirement("{} {}".format(name, specifier)),
+            user_supplied=any(r.user_supplied for r in ireqs),
+            comes_from=ireqs[0].comes_from,
+            use_pep517=any(r.use_pep517 for r in ireqs),
+            isolated=any(r.isolated for r in ireqs),
+            constraint=all(r.constraint for r in ireqs),
+            install_options=list(itertools.chain.from_iterable(
+                r.install_options for r in ireqs
+            )),
+            global_options=list(itertools.chain.from_iterable(
+                r.global_options for r in ireqs
+            )),
+            hash_options=dict(itertools.chain.from_iterable(
+                r.hash_options.items() for r in ireqs
+            )),
+        )
```

Comment:
Oh no.

Reply:
OMG, this sucks like a sucky thing that's just won the world sucking championships. If this passes the tests, I'll be impressed and disturbed in about equal measure.

I don't think there's any better way of doing this, but it's going to be a maintenance nightmare. I would honestly give serious consideration to *not* fixing #9020 rather than making this change. The issue is a consequence of the fact that we stripped constraints back to being "just" limits on what versions the resolver sees, rather than being full requirements with a weird "but don't install anything unless there's another requirement for the same package" limitation. I understand the OP's concerns in #9020 but at some point we *have* to start clamping down on the whole "it was never documented but I use it so you have to support it forever" thing - or we'll end up blocked on ever changing pip, and we'll have another distutils on our hands 🙁 

I'd characterise the change in behaviour as an "intentional regression", as the OP in #9020 put it.

Reply:
> OMG, this sucks like a sucky thing that's just won the world sucking championships.

lol

I wouldn't go that faaar, but yea, re-reading #9020 I think it's fine to put this into "constraints aren't the right tool for this" camp.

Reply:
> I wouldn't go that faaar

On re-reading, I should be clear, I mean the whole "`InstallRequirement` needs a template, and we don't have one so we have to make something up" situation, not @uranusjr's code (which as I say is about the best we can do in the situation). Sorry @uranusjr if it came across as criticising your code...

Reply:
Good news, there are regressions that I don’t understand and not very interested in looking into.